rpmlint/checks/ZipCheck.py

```diff
@@ -19,72 +19,82 @@ def check(self, pkg):
+
+                        # jar checks
                         if self.jar_regex.search(fname):
-                            if self._check_classpath(z):
-                                self.output.add_info('W', pkg, 'class-path-in-manifest', fname)
-                            if self._check_jarindex(z):
-                                self.output.add_info('W', pkg, 'jar-not-indexed', fname)
+                            self._check_classpath(pkg, fname, z)
+                            self._check_jarindex(pkg, fname, z)
                 except BadZipFile as err:
                     self.output.add_info('E', pkg, 'unable-to-read-zip', f'{fname}: {err}')
                 except RuntimeError as err:
                     self.output.add_info('W', pkg, 'unable-to-read-zip', f'{fname}: {err}')
 
-    @staticmethod
-    def _check_compression(zipfile):
+    def _check_bad_crc(self, pkg, fname, zipfile):
+        """
+        Check CRC issues for the files in the zipfile.
+
+        Print an error if there is a file in the archive that fails CRC check.
         """
-        Check if zip is actually compressed.
-        One file with smaller size is enough.
+        badcrc = zipfile.testzip()
+        if badcrc:
+            self.output.add_info('E', pkg, 'bad-crc-in-zip', badcrc, fname)
+
+    def _check_compression(self, pkg, fname, zipfile):
+        """
+        Check if zip is actually compressed or not.
+
+        One file with smaller size is enough. Print an error if the zipfile
+        is not compressed.
         """
-        # check for empty archives which are walid
+        # check for empty archives which are valid
         filecount = len(zipfile.namelist())
         nullcount = 0
         for zinfo in zipfile.infolist():
             if zinfo.file_size == 0:
                 nullcount += 1
             if zinfo.compress_size != zinfo.file_size:
-                return True
+                return
+
         # empty files only
         if filecount == nullcount:
-            return True
-        return False
+            return
+        self.output.add_info('E', pkg, 'uncompressed-zip', fname)
 
-    @staticmethod
-    def _check_classpath(zipfile):
+    def _check_classpath(self, pkg, fname, jarfile):
         """
-        Check if package contains MANIFEST.MF without classpath
+        Check if META-INF/MANIFEST.MF file in the jar contains a hardcoded
+        Class-Path.
+
+        Print a warning if the path is hardcoded.
         """
         classpath_regex = re.compile(r'^\s*Class-Path\s*:', re.MULTILINE | re.IGNORECASE)
 
+        # the META-INF is optional so skip if it is not present
         mf = 'META-INF/MANIFEST.MF'
-        # The META-INF is optional so skip if it is not present
-        if mf not in zipfile.namelist():
-            return False
-        # otherwise check for the classpath
-        manifest = zipfile.read(mf).decode()
+        if mf not in jarfile.namelist():
+            return
```

Comment:
As this function and the one bellow check for this meta-inf I guess this could even be detected prior going to these two functions.

Reply:
Actually this meta-inf check is not necessary for _check_jarindex() method so I removed it completely.